### ISSUE TITLE
:term should forward unknown sequences to host

### ISSUE THREAD

**Issue by chemzqm** [neovim/neovim#4349]
iTerm2 version 3 beta support show image inside terminal with command like `imgcat`, but the image could not be seen with that command inside terminal of neovim.
That would also be cool if neovim support show image in buffer with an api .

**Comment by justinmk** [comment]
Some people are already doing this, see https://github.com/neovim/neovim/issues/4056

**Comment by chemzqm** [comment]
@justinmk I'm using latest neovim, but I can't get any image ouput with command `imgcat` inside the terminal of neovim

**Comment by justinmk** [comment]
It looks like [imgcat is using](https://raw.githubusercontent.com/gnachman/iTerm2/master/tests/imgcat) a custom DCS `1337`. I am guessing that is not a code that is recognized by libvterm (@leonerd can you confirm), but on the other hand we should forward it to iTerm2, but maybe we aren't?

@chemzqm Clearly there is _some_ seequence that is recognized by libvterm to display images as demonstrated by https://github.com/neovim/neovim/issues/4349. imgcat could be modified to emit those codes instead?

**Comment by DarkDefender** [comment]
@justinmk IIRC the method in #4056 does not use terminal codes. It instead calls a image display program that displays the image at the requested location.

That's probably why it doesn't output the image at the correct location when inside neovim. Because it assumes that it is using the full terminal window. So it will not work with splits and so on.

I guess that the iTerm2 codes will eliminate this problem because the terminal will then know where to put the image depending on where it reads the DCS 1337 code

**Comment by sassanh** [comment]
+1

**Comment by rzpilgrim** [comment]
+1

**Comment by sassanh** [comment]
hi @justinmk 
sorry to ping you here. I just got a notif from github that my comment here has been deleted. Do you have any idea why? Deletion of my comment is not important by itself, I just found it suspicious and thought I should inform you, specially considering another comment has been deleted.
I don't know about contents of the comments but I'm sure mine aged at least 6 months.

**Comment by sassanh** [comment]
Checking my email I found the other comment was a "+1", unfortunately I don't have my own comment's content in my email but I guess it was also a "+1" and maybe "neovim" is a bot that deletes +1s which is good.

**Comment by liubog2008** [comment]
imgcat can work well in vim but in neovim it prints chars @justinmk

**Comment by liubog2008** [comment]
It looks that `vterm_state_set_unrecognised_fallbacks` is not called though libvterm is updated?

**Comment by kdheepak** [comment]
Is there any update to this issue? It would be nice to have this feature. 

I understand that this is related to forwarding unknown sequences to iTerm2. However, my understanding is that sixel support in libvterm is also an option for displaying images, right? I'm not familiar with the internals/architecture, but is it still the case that neovim's internal screen is composed of unicode characters of fixed size and hence will never be able to display images [1]? Is this unlikely to change?

<img width="1680" alt="Screen Shot 2019-06-20 at 1 30 22 AM" src="https://user-images.githubusercontent.com/1813121/59829628-206d6e80-92fb-11e9-90b0-1fe46d2188b4.png">

With iterm2 3.0.0 beta it is now possible to display sixel images as well [2]. In the screenshot above on the left I'm able to display images using `imgcat` (iTerm2 escape codes) and using `cat` (SIXEL). 

One valid use case for this feature is that these can be powerful tools for visualization / exploring data in scientific computing from the terminal interactively. It appears that tmux is not planning to support this [3], so neovim's terminal would make a highly appealing option to interactively run code and visualize data, all from the command line. 

It'd be really cool to have one or both methods supported in neovim. Thoughts?

[1] [https://github.com/neovim/neovim/issues/4500](https://github.com/neovim/neovim/issues/4500)
[2] https://gitlab.com/gnachman/iterm2/issues/3543#note_157693474
[3] https://github.com/tmux/tmux/issues/1613

**Comment by bfredl** [comment]
> but is it still the case that neovim's internal screen is composed of unicode characters of fixed size and hence will never be able to display images.

"never" is the wrong word, the question is if it is worth the effort or not. Does the sixel state of a single cell fit into 28 ASCII chars, or something like that? Then we could add support without changing the screen representation fundamentally.

**Comment by kdheepak** [comment]
I'm not sure what you mean by "sixel state" or "a single cell"? I'm also just learning about all of this, so take it for what it is worth. The following information is based on the vt100 reference manual [1]. 

A single "sixel" encodes information about a six pixel high * one pixel wide column. So each sixel in binary is encoded as a single 6-bit number. This is converted to ascii such that every sixel is within the printable character range of the ASCII character set. 

Wouldn't just passing on sixel information to iTerm2 do the trick?

[1] https://www.vt100.net/docs/vt3xx-gp/chapter14.html

**Comment by kdheepak** [comment]
Other terminals also appear to be using their own format for displaying images. Kitty for example also has similar escape sequences to iTerm2: https://sw.kovidgoyal.net/kitty/graphics-protocol.html

**Comment by bfredl** [comment]
> A single "sixel" encodes information about a six pixel high * one pixel wide column. So each sixel in binary is encoded as a single 6-bit number. This is converted to ascii such that every sixel is within the printable character range of the ASCII character set.

I understood so much. But how many sixels wide is a text cell defined to be? how many "sixel lines" per one text line? Or if this is not defined and varies among terminals/configs, can it be detected somehow?

> Wouldn't just passing on sixel information to iTerm2 do the trick?

To keep the vterm in a consistent state we need at least to know how large the sixel area is. But I think users expect if they scroll/rezise the terminal, or display it in a different window, the graphics will be retained, and thus we would like to store the state in our internal screen buffers.

**Comment by nall** [comment]
I’ll comment since #11282 was marked a dupe of this. I am not interested in images so much as other features iterm2 exposes via these codes.

Please consider passing these codes through even if there are questions about the image support since that’s only one of the ways they are used.

**Comment by bradparks** [comment]
Although dithered, you can use asciimatics to view a dithered image in a neovim terminal or tmux:

https://stackoverflow.com/a/59223021/26510

**Comment by ElPiloto** [comment]
Any progress or decisions made on this?  Adding my intended use case just in case that helps tip the scale:
- images (specifically graphs/plots) in neovim's terminal would go a long way to using neovim as a python IDE for machine learning.

**Comment by yatli** [comment]
Also curious about this. Could help to implement GUI elements for neovim.

**Comment by tejasvi** [comment]
[It is possible to display images](https://memotut.com/image-display-and-video-playback-on-neovim-terminal-8a82c/) by disabling embedded `libterm` during build with `-DUSE_BUNDLED_LIBVTERM=OFF`.

**Comment by bfredl** [comment]
it requires this patch tho https://github.com/arakiken/libvterm/commit/2500d7ad1c7f74162d70e3089fc95934044b8759

@leonerd does it look like something that could be included upstream?

**Comment by sassanh** [comment]
It is not working with iTerm2:
<img width="1351" alt="image" src="https://user-images.githubusercontent.com/1270688/107847468-2a61b080-6e01-11eb-85e2-c3bea14c9f87.png">

**Comment by bluss** [comment]
Image support would be awesome. It seems the issue title is a bit of a misstatement, thought: just forwarding unknown sequences to host would not make for great support, then scrollback, windows and views etc will have no chance of working correctly. It looks like libvterm and neovim need to support the image protocols for terminals explicitly (and unofficial libvterm patches for this are posted in earlier comments in this issue). Issue #4500 has a more specific target and is sixel-specific, but any image protocol would be a good start.

**Comment by starptr** [comment]
Has arakiken's patch been merged into upstream libvterm yet? There is a build of neovim at [nat-chan/Neovim-DRCS-SIXEL-dtach](https://github.com/nat-chan/Neovim-DRCS-SIXEL-dtach) that uses that patch and shows sixels working

**Comment by neovim-discourse** [comment]
This issue has been mentioned on **Neovim Discourse**. There might be relevant details there:

https://neovim.discourse.group/t/migrating-libvterm-to-somewhere-to-attract-more-contribution/3100/1

**Comment by leonerd** [comment]
"Oops! That page doesn’t exist or is private." - I can't see it.

**Comment by clason** [comment]
It's been deleted.

**Comment by leonerd** [comment]
What's the overall status here then? The original subject is about forwarding unrecognised sequences - I'm fairly sure libvterm already has facility to do that now. Is there something still missing in it? If so let me know and I can add it.

If not, and that ability is there now perhaps we can close this issue, in favour of a different one discussing particular things to be handled..?

**Comment by sassanh** [comment]
Thanks for the update!
Are these facilities documented somewhere? How can I learn about it?

**Comment by leonerd** [comment]
> Are these facilities documented somewhere? How can I learn about it?

The `VTermStateFallbacks` structure in include/vterm.h

**Comment by justinmk** [comment]
The typical use case mentioned here was `imgcat`. Assuming that works now, we can close this. Thanks @leonerd

**Comment by jaekyeom** [comment]
@justinmk I tried `imgcat` in terminal buffers with both v0.7.2 and the [nightly](https://github.com/neovim/neovim/commit/12fe197cffc687cedc65259d9fd8351a25aa7792) build, but I think it still doesn't work.

**Comment by sassanh** [comment]
I registered this `osc` callback:

```c
static int oscCallback(const char *c, size_t s, void *a)
{
  int fd;
  fd = open("/tmp/log", O_WRONLY | O_CREAT | O_APPEND, 0644);

  write(fd, "\e]", 2);
  write(fd, c, s);
  write(fd, "\7", 1);

  close(fd);
  return 1;
}

VTermParserCallbacks callbacks = {
  .osc = oscCallback,
};
...
  vterm_state_set_unrecognised_fallbacks(state, &callbacks, NULL);
```
then I was able to do these in Neovim terminal and see the results in another normal iTerm2 terminal by running `cat /tmp/log` in it:
1. Cursor shape changes
2. Changing iTerm 2 title

Everything worked as expected when I cat `/tmp/log` except `imgcat`, orc callback is not triggered when `imgcat` runs. Seems like `libvterm` is doing its job. But when instead of `fd` I wrote to stdout, things got ugly and part of the prompt rendered out of the terminal window in another buffer. I guess in Neovim realm, communicating with the underlying terminal Is not as easy as directly writing to stdout.

We may be able to get rid of the virtual cursor and let the running application (bash, fish, node, etc) directly set the cursor shape by communicating with the underlying terminal emulator.

Also it may break some people's workflow, commands that were ignored in the past, now reach the underlying terminal emulator, there may be resize commands, setting title commands, setting working directory commands, etc so it will be a breaking change (someone may have an iTerm2 script that does something based on the title of the terminal and doesn't expect Neovim's embedded terminal change that, etc) so we may want to start with a feature flag that is disabled by default.

Note for @leonerd: `imgcat` did not work as none of the fallback callbacks were called after running `imgcat`, I'm going to read the code to find out what are the conditions for calling `osc` callback and which condition is not satisfied with `imgcat` output. Please let me know if you have any suggestions.

**Comment by leonerd** [comment]
I have no idea what `imgcat` is. Is it supposed to be outputting OSCs? I'll need some more reference to understand all this.

**Comment by sassanh** [comment]
Sure, this is the script: https://iterm2.com/utilities/imgcat
Yes it is supposed to be outputting OSCs, I attach a sample output [here](https://gist.github.com/sassanh/b870b1dfb708be0d37662b8d2d2b6f5c#file-sampel_imgcat_output-osc)

**Comment by sassanh** [comment]
I ran this command in a terminal opened by `:term` in Neovim:
```bash
imgcat "<IMAGE_PATH>" > /tmp/sampe;_imgcat_output.osc
```

**Comment by leonerd** [comment]
Ah wait a second - the `oscCallback` function you defined earlier isn't even compatible with the required signature. The fallbacks structure requires:

```c
  int (*osc)(int command, VTermStringFragment frag, void *user);
```

following the VTermStringFragment API. Perhaps that's the problem here?

**Comment by sassanh** [comment]
That's not what I see I the version of vterm used in Neovim:
```c
typedef struct {
  int (*text)(const char *bytes, size_t len, void *user);
  int (*control)(unsigned char control, void *user);
  int (*escape)(const char *bytes, size_t len, void *user);
  int (*csi)(const char *leader, const long args[], int argcount, const char *intermed, char command, void *user);
  int (*osc)(const char *command, size_t cmdlen, void *user);
  int (*dcs)(const char *command, size_t cmdlen, void *user);
  int (*resize)(int rows, int cols, void *user);
} VTermParserCallbacks;
```

Also it is already working for cursor shape commands, I checked it byte by byte. Are we using an old version of vterm in Neovim?

**Comment by leonerd** [comment]
Wow that seems pretty old. Yup that's probably what's causing it. The old version without the streaming-string API had trouble with long string escapes - such as is probably what the image mechanism uses. If you update to a more recent version you'll most likely find it nicely working.

**Comment by sassanh** [comment]
So we should be in a better position after [this PR](https://github.com/neovim/neovim/pull/17329) is merged.

**Comment by leonerd** [comment]
That PR is the reflow branch. You don't want that yet while it's still not actually working (at all. It crashes in simple situations). You just want an updated libvterm mainline.

**Comment by sassanh** [comment]
Thanks @leonerd! Yes it worked:

I applied this diff:
```diff
diff --git a/cmake.deps/CMakeLists.txt b/cmake.deps/CMakeLists.txt
index d2047c5b3..084b654f7 100644
--- a/cmake.deps/CMakeLists.txt
+++ b/cmake.deps/CMakeLists.txt
@@ -177,8 +177,8 @@ set(UNIBILIUM_SHA256 29815283c654277ef77a3adcc8840db79ddbb20a0f0b0c8f648bd8cd49a
 set(LIBTERMKEY_URL https://www.leonerd.org.uk/code/libtermkey/libtermkey-0.22.tar.gz)
 set(LIBTERMKEY_SHA256 6945bd3c4aaa83da83d80a045c5563da4edd7d0374c62c0d35aec09eb3014600)
 
-set(LIBVTERM_URL https://www.leonerd.org.uk/code/libvterm/libvterm-0.1.4.tar.gz)
-set(LIBVTERM_SHA256 bc70349e95559c667672fc8c55b9527d9db9ada0fb80a3beda533418d782d3dd)
+set(LIBVTERM_URL https://www.leonerd.org.uk/code/libvterm/libvterm-0.2.tar.gz)
+set(LIBVTERM_SHA256 4c5150655438cfb8c57e7bd133041140857eb04defd0e544521c0e469258e105)
 
 set(LUV_VERSION 1.44.2-0)
 set(LUV_URL https://github.com/luvit/luv/archive/1.44.2-0.tar.gz)
diff --git a/src/nvim/terminal.c b/src/nvim/terminal.c
index 90966bcfa..fade76c97 100644
--- a/src/nvim/terminal.c
+++ b/src/nvim/terminal.c
@@ -154,6 +154,31 @@ static VTermScreenCallbacks vterm_screen_callbacks = {
   .sb_popline  = term_sb_pop,
 };
 
+static int osc(int command, VTermStringFragment fragment, void *user)
+{
+  int fd;
+  fd = open("/tmp/osc", O_WRONLY | O_CREAT | O_APPEND, 0644);
+
+  if (fragment.initial) {
+    write(fd, "\e]", 2);
+    char s[100];
+    size_t length = (size_t)sprintf(s, "%d;", command);
+    write(fd, s, length);
+  }
+  write(fd, fragment.str, fragment.len);
+  if (fragment.final) {
+    write(fd, "\7", 1);
+  }
+
+  close(fd);
+  return 1;
+}
+
+VTermStateFallbacks callbacks = {
+  .osc = osc,
+};
+
+
 static PMap(ptr_t) invalidated_terminals = MAP_INIT;
 
 void terminal_init(void)
@@ -275,6 +300,8 @@ Terminal *terminal_open(buf_T *buf, TerminalOptions opts)
     }
   }
 
+  vterm_state_set_unrecognised_fallbacks(state, &callbacks, NULL);
+
   return rv;
 }
 
@@ -886,8 +913,8 @@ static int term_settermprop(VTermProp prop, VTermValue *val, void *data)
     break;
 
   case VTERM_PROP_TITLE: {
-    buf_T *buf = handle_get_buffer(term->buf_handle);
-    buf_set_term_title(buf, val->string);
+    // buf_T *buf = handle_get_buffer(term->buf_handle);
+    // buf_set_term_title(buf, val->string);
     break;
   }
 
```

Then saved this JS code as `debounce.js`:
```javascript
let buffer = Buffer.alloc(0);

function flush() {
  process.stdout.write(buffer);
  buffer.fill();
}

let id = setInterval(flush, 1000);

process.stdin.on('data', data => {
  clearInterval(id);
  buffer = Buffer.concat([buffer, data]);
  id = setInterval(flush, 1000);
});
```

Then ran this command in a normal iTerm2 tab:
```sh
tail -f /tmp/osc | node debounce.js
```

And running `imgcat` in a terminal inside Neovim I could see it in the iTerm2 terminal.

<img width="1208" alt="image" src="https://user-images.githubusercontent.com/1270688/188531064-4b029e25-5f70-44e1-95d8-0e9e648c780a.png">

The debouncer is needed so that the whole imgcat command is sent to iTerm2 in a single write, it wouldn't work otherwise. So we will need to buffer fragments and concatenate them in a buffer before writing to terminal.

When I hidden the buffer and opened it again, it didn't call the osc callback again (osc callback is called only once). I think this should be handled in Neovim and not libvterm and I think it won't be easy because some osc commands (like showing a notification, bouncing iTerm2 icon, etc) should run only once even if the buffer is hidden and shown again, but some like the one generated by `imgcat`, should run again exactly in the same place whenever the buffer is shown again. Deciding which osc command is in which category may need collaboration with iTerm2 developer for a general and future proof solution, alternatively we can hardcode the list of the commands in the second category in Neovim's codebase.

**Comment by justinmk** [comment]
@sassanh nice, I don't see a PR for that so would be worth getting one started.

**Comment by justinmk** [comment]
see https://github.com/neovim/neovim/pull/17329

**Comment by sassanh** [comment]
@justinmk As Paul [mentioned](https://github.com/neovim/neovim/issues/4349#issuecomment-1237543342) that PR is for the reflow branch which is work in progress and buggy. Upgrading libvterm to 0.2 is much easier than that. I will try to create a pull request.

I'm sorry my previous comment is too long, but it would be nice if someone familiar with the codebase reads it and share their ideas about the challenges I stated.

**Comment by clason** [comment]
@sassanh No, we tried that before, and that version has issues which cause our tests to fail. The decision has been made to therefore skip 0.2 and update to the next version that is based on the reflow branch (which doesn't have _these_ issues).

**Comment by leonerd** [comment]
> The decision has been made to therefore skip 0.2 and update to the next version that is based on the reflow branch (which doesn't have _these_ issues).

Oh, nobody told me that. I haven't been *actively* working on reflow. Without knowing that you'd have been waiting indefinitely. It seems we're stuck at a stalemate then. I really would suggest *not* waiting on reflow because it is tricky and buggy and nobody seems to be helping me fix it, so it could take years.

**Comment by sassanh** [comment]
@clason Reading the comments I could not understand what is the bug in version 0.2 of libvterm. Are you aware of it? Do you have a reference to the chat you mentioned [here](https://github.com/neovim/neovim/pull/17329#issuecomment-1116108418)? I'm hoping maybe we can fix the bug in a 0.2.1 version and upgrade libvterm without being blocked by reflow stuff.
I'm indeed interested in reflow too, I'm just hoping maybe we can separate these things and resolve them one by one.

**Comment by clason** [comment]
No, I don't know anything beyond the comments in these two PRs.

If you can fix this and help get a 0.2.1 release in the next two weeks, we can include this for 0.8; otherwise we'd prefer to take the time and target reflow for 0.9.

**Comment by tmke8** [comment]
Now that libvterm has been updated to 0.3.3 (#24547), is this now easy to implement? Should @sassanh 's [diff posted above](https://github.com/neovim/neovim/issues/4349#issuecomment-1237579330) just work?

**Comment by clason** [comment]
Why don't you try it? In any case, we now use an unpatched libvterm and want to keep it this way. So any changes would have to be made on the libvterm side, not here. If there's a libvterm version which support this, we'll make use of it of course.

**Comment by tmke8** [comment]
Okay, I applied the patch on neovim master and it works, but I didn't realize beforehand that it requires running something in another terminal window and that the image would only appear in that other window. I'm assuming that making the image appear within neovim, in the built-in terminal, is much harder than this.

**Comment by justinmk** [comment]
> It looks like [imgcat is using](https://iterm2.com/utilities/imgcat) a custom DCS `1337`.

OSC/DSC are now forwarded to the `TermRequest` autocmd event. See also https://github.com/neovim/neovim/pull/28621 .

**Comment by DUOLabs333** [comment]
@justinmk How would one write a handler that forwards this to the host, or does/should the default handler do that automatically?

**Comment by justinmk** [comment]
> How would one write a handler that forwards this to the host,

Handle `TermRequest` and write to `io.stdout:write()`. See example at `TermResponse` for a skeleton.
